### PR TITLE
[FEATURE] Add documentation validator

### PR DESCRIPTION
**Pull request by alexander-nitsche**
The documentation validator checks the documentation's reST files and Settings.cfg for compliance with the recently introduced TYPO3 documentation standards [1][2] and lists selected errors. The authors identified by composer.json are notified and the Intercept history log provides the validation errors for the remaining contributors.

For now, the validation errors are handled gently and documentation rendering is performed anyway. After a sufficient grace period, strict validation will be introduced.

[1] https://docs.typo3.org/m/typo3/docs-how-to-document/main/en-us/GeneralConventions/FileStructure.html
[2] https://docs.typo3.org/m/typo3/docs-how-to-document/main/en-us/UserRoundTrip.html

### CODE REVIEW

**Comment by alexander-nitsche** [comment]
This is how the new email format looks like:
Screen Shot 2022-05-07 at 00 46 36

**Comment by alexander-nitsche** [comment]
This is how the history log entry of a successful documentation rendering with a validation warnings looks like:
Screen Shot 2022-05-07 at 00 49 15

**Comment by alexander-nitsche** [comment]
Slightly optimized the email to:
Screen Shot 2022-05-07 at 09 53 30

File: src/Service/MailService.php [hunk 1]
```diff
@@ -29,35 +29,80 @@ public function __construct(Swift_Mailer $mailer, Environment $templating)
...
+                        'author' => $firstAuthor,
+                        'package' => $composerJson->getName(),
+                        'pushEvent' => $pushEvent,
+                        'reasonPhrase' => $warningMessage,
+                    ]
+                );
+                $message
+                    ->setFrom('intercept@typo3.com')
+                    ->setTo($firstAuthor['email'])
+                    ->setCc(array_map(function ($author) {
```
Comment (NeoBlack) [inline src/Service/MailService.php:58]: Why CC and not TO?

File: src/Controller/DocsRenderingController.php [hunk 2]
```diff
@@ -97,10 +102,20 @@ public function index(
                         $manager->flush();
                         continue;
                     }
+                    $historyEntryAdditionalData = [];
                     $composerJson = $documentationBuildInformationService->fetchRemoteComposerJson($pushEvent->getUrlToComposerFile());
                     $composerAsObject = $documentationBuildInformationService->getComposerJsonObject($composerJson);
                     $buildInformation = $documentationBuildInformationService->generateBuildInformation($pushEvent, $composerAsObject);
                     $documentationBuildInformationService->assertBuildWasTriggeredByRepositoryOwner($buildInformation);
+                    // Handle documentation validation errors gently for a grace period. Remove this try-catch wrap
```
[the remark this answers is not in the document]
Comment (alexander-nitsche) [reply src/Controller/DocsRenderingController.php:110]: The idea of the grace period is to collect feedback and improve the validator in production. I assume that there will be some feedback as soon as authors are publishing new documentation versions, so i would tend to have a longer grace period, maybe half a year. The is no deadline i would commit to for now.

File: src/Service/MailService.php (hunk 1 above)
Comment (alexander-nitsche) [reply src/Service/MailService.php:58]: This is for historic reasons: The first author mentioned in the composer.json was the only recipient before this patch. now (s)he still is special by being to only one in TO, and the rest is in CC. This was only a stylistic reason, no fundamental one.

File: src/Service/MailService.php [hunk 3]
```diff
@@ -29,35 +29,80 @@ public function __construct(Swift_Mailer $mailer, Environment $templating)
...
+                        'package' => $composerJson->getName(),
+                        'pushEvent' => $pushEvent,
+                        'reasonPhrase' => $exceptionMessage,
+                    ]
+                );
                 $message
                     ->setFrom('intercept@typo3.com')
-                    ->setTo($composerJson->getFirstAuthor()['email']);
+                    ->setTo($firstAuthor['email'])
+                    ->setCc(array_map(function ($author) {
```
[the remark this answers is not in the document]
Comment (alexander-nitsche) [reply src/Service/MailService.php:98]: See above.

File: src/Extractor/ComposerJson.php [hunk 4]
```diff
@@ -70,10 +70,12 @@ public function getCoreRequirement(): ?string
      * @return array
      * @throws DocsComposerMissingValueException
      */
-    public function getFirstAuthor(): array
+    public function getAuthorsHavingEmailAddress(): array
     {
         $this->assertPropertyContainsValue('authors');
-        return current($this->composerJson['authors']);
+        return array_filter($this->composerJson['authors'], function ($author) {
+            return filter_var($author['email'] ?? '', FILTER_VALIDATE_EMAIL);
```
[the remark this answers is not in the document]
Comment (alexander-nitsche) [reply src/Extractor/ComposerJson.php:77]: This filter was in use before this patch already. Furthermore it is used also in other places of this project already, see e.g.
```
vendor/phar-io/manifest/src/values/Email.php
vendor/phpdocumentor/reflection-docblock/src/DocBlock/Tags/Author.php
vendor/symfony/maker-bundle/src/Validator.php
vendor/webmozart/assert/src/Assert.php
```

I think this is not a critical part that must be 100% correct, as we are not sending critical data in the email, but if you would like to switch to a different validation, please tell me which. The \Egulias\EmailValidator\EmailValidator?

**Review by linawolf — CHANGES_REQUESTED** [review]
We do not have strict TYPO3 Documentation standards. We merely have guidelines that we do not intent to enforce. So I as Co-Team-Lead do not agree with enforcing these guidelines. Therefore this PR is not needed and contra-productive as it would just cause much work to many extension owners without any clear benefit.

**Comment by alexander-nitsche** [comment]
We have standards and any check of documentation, that can be performed automatically, should be performed automatically. There is no benefit in sticking to manual checks or no checks at all. Automatic checks help being productive by allowing to concentrate on different topics.

**Comment by linawolf** [comment]
So, I asked Martin about it and we both decided that this is neither needed nor wanted. @NeoBlack please close.

**Comment by NeoBlack** [comment]
As decided by the documentation team leaders, I close this PR.

**Pull request closed without merging**